python/pyspark/sql/group.py

```diff
@@ -233,6 +233,27 @@ def apply(self, udf):
         |  2| 1.1094003924504583|
         +---+-------------------+
 
+        Notes on grouping column:
```

Comment:
This explains the general idea. I plan to improve the doc if people think this change is good.

Reply:
It looks more reasonable to me to pass the grouping columns to UDF and let the UDF to decide if it wants to include the grouping columns or not.

Reply:
+1 for ^

Reply:
@cloud-fan That's what I thought too initially. Let's consider this use case,

```
import statsmodels.api as sm
# df has four columns: id, y, x1, x2

group_column = 'id'
y_column = 'y'
x_columns = ['x1', 'x2']
schema = df.select(group_column, *x_columns).schema

@pandas_udf(schema, PandasUDFType.GROUP_MAP)
# Input/output are both a pandas.DataFrame
def ols(pdf):
    group_key = pdf[group_column].iloc[0]
    y = pdf[y_column]
    X = pdf[x_columns]
      X = sm.add_constant(X)
    model = sm.OLS(y, X).fit()

    return pd.DataFrame([[group_key] + [model.params[i] for i in   x_columns]], columns=[group_column] + x_columns)

beta = df.groupby(group_column).apply(ols)
```
This is a simple pandas UDF that does a linear regression. The issue is, although the UDF (linear regression) has nothing to do with the grouping column, the user needs to deal with grouping column in the UDF. In other words, the UDF is coupled with the grouping column.

If we make it such that grouping columns are prepend to UDF result, then the user can write something like this:

```
import statsmodels.api as sm
# df has four columns: id, y, x1, x2

group_column = 'id'
y_column = 'y'
x_columns = ['x1', 'x2']
schema = df.select(*x_columns).schema

@pandas_udf(schema, PandasUDFType.GROUP_MAP)
# Input/output are both a pandas.DataFrame
def ols(pdf):
    y = pdf[y_column]
    X = pdf[x_columns]
      X = sm.add_constant(X)
    model = sm.OLS(y, X).fit()

    return pd.DataFrame([[model.params[i] for i in   x_columns]], columns=x_columns)

beta = df.groupby(group_column).apply(ols)
```

Now the UDF is cleaner because it only deals with columns that are relevant to the regression. It also make the UDF more reusable, as the user can now do something like:

```
beta1 = df.groupby('a').apply(ols)
beta2 = df.groupby('a', 'b').apply(ols)
```
Because the UDF is now decoupled with the grouping column, the user can reuse the same udf with different grouping, which is not possible with the current API.

@cloud-fan @HyukjinKwon What do you think?

Reply:
Yup, I saw this usecase as described in the JIRA and I got that the specific case can be simplified; however, I am not sure if it's straightforward to the end users.

For example, if I use `pandas_udf` I think I would simply expect the return schema is matched as described in `returnType`. I think `pandas_udf` already need some background and I think we should make it simpler as possible as we can.

It might be convenient to make the guarantee on grouping columns in some cases vs this might be a kind of magic inside.

I would prefer to let the UDF to specify the grouping columns to make this more straightforward more ..

Reply:
From a SQL background, I think we should add grouping keys to the input of UDF. Sometimes users do need to read the grouping keys when aggregating, and we should give users a way to do it. BTW this is also consistent with Dataset, see `KeyValueGroupedDataset.mapGroups`.

Reply:
@cloud-fan Good point. I did some more research on `KeyValueGroupedDataset. flatMapGroups ` and also `gapply` in SparkR. These two seem to have consistent behavior:

* Grouping key is passed as the first arg to the udf, the second arg is the actual data
* No grouping key is prepend to the result, i.e., the output of the udf is the schema of the result Spark DataFrame.

I think this could be a good option for `groupby apply` too for the following reason:
* Consistent with similar API in Spark, i.e., `flatMapGroups` and `gapply`
* Is flexible enough to implement the two use cases above:
```
df = ... # id int, v double
pandas_udf('id int, v double', GROUP_MAP)
def foo(key, pdf):
      # key is not used
      return pdf.assign(v=pdf.v+1)

df.groupby('id').apply(foo)
```
and
```
import statsmodels.api as sm
# df has four columns: id, y, x1, x2

group_column = 'id'
y_column = 'y'
x_columns = ['x1', 'x2']
schema = df.select(group_column, *x_columns).schema

@pandas_udf(schema, PandasUDFType.GROUP_MAP)
# Input/output are both a pandas.DataFrame
def ols(key, pdf):
    y = pdf[y_column]
    X = pdf[x_columns]
    X = sm.add_constant(X)
    model = sm.OLS(y, X).fit()

    return pd.DataFrame([key + [model.params[i] for i in x_columns]])

beta = df.groupby(group_column).apply(ols)
```
In the second example here, output schema for the udf is still coupled with grouping key, but the function itself is decoupled.

The downside is it's now different from the pandas groupby apply API (the pandas one only takes one argument), but we cannot be consistent with everything...

Thoughts?

Reply:
One thing probably needs a bit clarify that I am a little confused by the doc of `gapply` here:

https://github.com/apache/spark/blob/master/R/pkg/R/DataFrame.R#L1626

In the output, there is a "Model" column:
https://github.com/apache/spark/blob/master/R/pkg/R/DataFrame.R#L1646

but it's not specified in the schema:
https://github.com/apache/spark/blob/master/R/pkg/R/DataFrame.R#L1631

Looks like it might be a mistake in the doc from the description of the function, but I am not sure.

Reply:
@felixcheung, WDYT?

To cut the context short, it's a Pandas map group API like `gapply` (not Pandas scalar udf).

Its current implementation is as follows

```python
def foo(pdf):
    pdf  # this is the Pandas DataFrame

pudf = pandas_udf(f=foo, returnType="id int, v double", functionType=GROUP_MAP)
df.groupby(group_column).apply(pudf)
```

First `'id int, v double'` describes the output schema and input `pdf` is the grouped Pandas's DataFrame.

As @icexelloss described above as a new proposal, looking at `gapply` in R at a glance again, seems making sense that we do:

```python
def foo(key, pdf):
    key  # this is a grouping key. 
    pdf  # this is the Pandas DataFrame

pudf = pandas_udf(f=foo, returnType="id int, v double", functionType=GROUP_MAP)
df.groupby(group_column).apply(pudf)
```

Reply:
sounds to me like we could either stick with func(key, pdf) or whatever pandas does.

(yes, for gapply, the returned data frame is expected to have key columns prepended; there was one SPARK-16258 proposing to eliminate that extra work)

Reply:
Up to my knowledge, the current implementation follows Pandas's default groupBy - apply when Pandas DataFrame -> Pandas DataFrame (correct me if I am wrong). So, I was thinking that we shouldn't start with prepending the grouping columns but we could alternatively consider an idea of `gapply` in somehow ..

I think it's still feasible to have both ideas - If the given function takes single argument, we can give the input as pdf. If it takes two arguments, we can give key and pdf as input. I think we can support the `gapply`-like support optionally.

It's a rough idea but I think we can do this in theory as we know and can `inspect` the function ahead before computation.

WDYT guys?

Reply:
SGTM, one thing is how to define the type of `key`, a row?

Reply:
Sorry for the late reply. I agree with @HyukjinKwon, I think we can do support both `foo(pdf)` and `foo(key, pdf)` through inspection.

I will try to put up a PR soon.

As to how to represent key, I think a tuple might be enough but I think a row also works. What do you guys think?

Reply:
It's interesting to see the discussion in `SPARK-16258`. I think this is quite hard for the API to meet all cases...But the `foo(key, pdf)` is the best so far I think.

Reply:
tuple sounds good, let's go with that.

Reply:
@HyukjinKwon's proposal sounds good to me too.

Reply:
I'd like to confirm what the result schema will be like finally.
If users want to include the keys, the udf should include the keys in its output and the keys will not be prepended automatically?

Reply:
yup

Reply:
I see, sounds good. Thanks!

Reply:
Thanks all for the discussion. I will update the Jira and open a new PR.